File: nodejs/default_tags.js
var sys = require('sys');
var extend = require('./util');
var Thistle = require('./thistle');
var node   = require('./node');

var CommentNode = function() { }

extend(CommentNode, node.Node, {
    render : function (context) { return ''; },
    toString : function() { return "<Comment Node>"; }
});

//
//
//
var AutoEscapeControlNode = function(setting, nodelist) { 
    this.setting = setting;
    this.nodelist = nodelist;
}

extend(AutoEscapeControlNode, node.Node, {
    render : function (context) { 
        var old_setting = context.autoescape;
        context.autoescape = this.setting;
        output = this.nodelist.render(context);
        context.autoescape = old_setting;
        if (this.setting)
            return new Thistle.mark_safe(output);
        return output;
    },
    toString : function() { return "<AutoEscapeControl Node>"; }
});

//
//
//
var CycleNode = function(cyclevars, variable_name, silent) { 
    this.cyclevars = cyclevars;
    this.variable_name = variable_name;
    this.silent = silent;
    this.ident  = 'CycleNode' + CycleNode.index++;
    this.pos = 0;
}
CycleNode.index = 100;

extend(CycleNode, node.Node, {
    render : function (context) { 
        if (typeof context.render_context[this.ident] == 'undefined')  {
            var self = this;
            context.render_context[this.ident] = function() {
                if (self.pos == self.cyclevars.length)
                    self.pos = 0;
                return self.cyclevars[self.pos++];
            }
        }
        cycle_iter = context.render_context[this.ident];
        value = cycle_iter().resolve(context);
        if (this.variable_name)
            context.set(this.variable_name, value);
        if (this.silent)
            return '';
        return value;
    },
    toString : function() { return "<Cycle Node>"; }
});

//
//
//
var FilterNode = function(filter_expr, nodelist) { 
    this.filter_expr = filter_expr;
    this.nodelist    = nodelist;
}

extend(FilterNode, node.Node, {
    render : function (context) { 
        var output = this.nodelist.render(context);
        // Apply filters
        context.update({'var':output});
        filtered = this.filter_expr.resolve(context);
        context.pop();
        return filtered;
    },
    toString : function() { return "<Filter Node>"; }
});

//
//
//
var ForNode = function(loopvars, sequence, is_reversed, nodelist_loop, nodelist_empty) {
    this.child_nodelists = ['nodelist_loop', 'nodelist_empty']

    this.loopvars = loopvars;
    this.sequence = sequence;
    this.is_reversed = is_reversed;
    this.nodelist_loop = nodelist_loop;
    if (nodelist_empty == null || nodelist_empty == undefined) 
        this.nodelist_empty = new node.NodeList();
    else
        this.nodelist_empty = nodelist_empty;
}

extend(ForNode, node.Node, {
    render : function(context) {
        var  parentloop;

        if (context.has_key('forloop')) {
            parentloop = context['forloop'];
        } else {
            parentloop = {};
        }

        context.push();

        var values = null;
        try {
            values = this.sequence.resolve(context, true);
        } catch(e) {
        }

        if (values == null)
            values = [];

        if (values.length < 1) {
            context.pop();
            return this.nodelist_empty.render(context);
        }

        var nodelist = new node.NodeList();
        if (this.is_reversed)
            values = values.reverse();

        var unpack = this.loopvars.length > 1;
        var loop_dict = { parentloop : parentloop };

        // Create a forloop value in the context.  We'll update counters on each
        // iteration just below.
        context.set('forloop', loop_dict);

        for (var i = 0; i < values.length; i++) {
            var  item = values[i];

            // sys.puts("item["+i+"] = " + item);

            loop_dict['counter0'] = i;
            loop_dict['counter'] = i+1;
            // Reverse counter iteration numbers.
            loop_dict['revcounter'] = values.length - i;
            loop_dict['revcounter0'] = values.length - i - 1;
            // Boolean values designating first and last times through loop.
            loop_dict['first'] = (i == 0);
            loop_dict['last'] = (i == values.length - 1);

            pop_context = false;

            if (unpack) {
                try {
                    unpacked_vars = {};
                    for (var idx = 0; idx < this.loopvars.length; idx++) {
                        unpacked_vars[this.loopvars[idx]] = item[idx];
                    }
                    pop_context = true;
                    context.update(unpacked_vars);
                } catch (e) {
                    // sys.puts(e);
                }
            } else {
                context.set(this.loopvars[0], item);
            }

            for (var idx = 0; idx < this.nodelist_loop.length; idx++) {
                nodelist.push(this.nodelist_loop[idx].render(context));
            }

            if (pop_context)
                context.pop();
        }

        context.pop();

        return nodelist.render(context);
    },
    toString : function() { return "<For Node>"; }
});

var IfNode = function(val, nodelist_true, nodelist_false) {
    this.nodelist_true = nodelist_true;
    this.nodelist_false = nodelist_false;
    this.val = val;
    this.child_nodelists = ['nodelist_true', 'nodelist_false'];
}
extend(IfNode, node.Node, {
    render: function(context) {
        try {
            val = this.val.eval(context);
        } catch(e) {
            val = null;
        }

        if (val) 
            return this.nodelist_true.render(context);
        return this.nodelist_false.render(context);
    },

    toString: function() {
        return "<IfNode>";
    },
});

//
//
//
var smartif = require('./smartif');

function TemplateLiteral(value, text) {
    smartif.Literal.call(this, value);
    this.text = text;
}
extend(TemplateLiteral, smartif.Literal, {
    display: function() {
        return this.text;
    },
    eval: function(context) {
        return this.value.resolve(context, true);
    },
});

function TemplateIfParser(parser, tokens) {
    this.error_class = Thistle.TemplateSyntaxError
    this.template_parser = parser;
    smartif.IfParser.call(this, arguments[1]);
}
extend(TemplateIfParser, smartif.IfParser, {
    create_var: function(value) {
        return new TemplateLiteral(this.template_parser.compile_filter(value), value);
    },
});

//
//
//
var DefaultTags = {
    comment : function(parser, token) {
        parser.skip_past('endcomment');
        return new CommentNode();
    },

    cycle : function(parser, token) {
        var args = token.split_contents()

        if (args.length < 2)
            throw new Thistle.TemplateSyntaxError("'cycle' tag requires at least two arguments");

        if (args[1].indexOf(',') != -1) {
            // Backwards compatibility: {% cycle a,b %} or {% cycle a,b as foo %} case.
            // sys.puts("ARGS = "+ args);
            // sys.puts("ARGS SPLIT = "+ args[1].split(','));
            args = [args[0]].concat(args[1].split(',').map(function (v) { return '"' + v + '"'; }), args.slice(2));
        }

        var name;

        if (args.length == 2) {
            // {% cycle foo %} case.
            name = args[1];

            if (typeof parser._namedCycleNodes == 'undefined')
                throw new Thistle.TemplateSyntaxError("No named cycles in template. '"+name+"' is not defined");;
            if (typeof parser._namedCycleNodes[name] == 'undefined')
                throw new Thistle.TemplateSyntaxError("Named cycle '"+name+"' does not exist");
            return parser._namedCycleNodes[name];
        }

        var as_form = false;
        var silent  = false;

        if (args.length > 4) {
            // {% cycle ... as foo [silent] %} case.
            if (args[args.length-3] == "as") {
                if (args[args.length-1] != "silent") {
                    throw new Thistle.TemplateSyntaxError("Only 'silent' flag is allowed after cycle's name, not '%s'." % args[-1]);
                }
                as_form = true;
                silent = true;
                args = args.slice(0, args.length-1);
            } else if (args[args.length-2] == "as") {
                as_form = true;
            }
        }

        var node = null;

        if (as_form) {
            name = args[args.length-1];
            var values = args.slice(1, args.length-2).map(function (arg) { return parser.compile_filter(arg); })
            node = new CycleNode(values, name, silent=silent);
            if (typeof parser._namedCycleNodes == 'undefined') 
                parser._namedCycleNodes = {}
            parser._namedCycleNodes[name] = node;
        } else {
            var values = args.slice(1).map(function (arg) { return parser.compile_filter(arg); })
            node = new CycleNode(values, null, false);
        }

        return node;
    },

    'for' : function(parser, token) {
        var bits = token.contents.split(/\s+/)

        if (bits.length < 4)
            throw new Thistle.TemplateSyntaxError("'for' statements should have at least four words: " + token.contents);

        var is_reversed = (bits[bits.length - 1] == 'reversed');
        var in_index = bits.length + (is_reversed ? -3 : -2);

        if (bits[in_index] != 'in')
            throw new Thistle.TemplateSyntaxError("'for' statements should use the format 'for x in y': " + token.contents);

        var loopvars = bits.slice(1, in_index).join(' ').split(/ *, */);
        for (var idx in loopvars) {
            var v = loopvars[idx];
            if (v.length == 0 || v.indexOf(' ') != -1)
                throw new Thistle.TemplateSyntaxError("'for' tag received an invalid argument: " + token.contents);
        }

        var sequence = parser.compile_filter(bits[in_index+1]);
        var nodelist_loop = parser.parse(['empty', 'endfor']);
        var nodelist_empty = null;

        token = parser.next_token();

        if (token.contents == 'empty') {
            nodelist_empty = parser.parse(['endfor'])
            parser.delete_first_token()
        }

        return new ForNode(loopvars, sequence, is_reversed, nodelist_loop, nodelist_empty);
    },

    'if' : function(parser, token) {
        var nodelist_true, nodelist_false;

        var bits = token.split_contents().slice(1);
        var p = new TemplateIfParser(parser, bits);
        var val = p.parse();

        nodelist_true = parser.parse(['else','endif']);
        token = parser.next_token();
        if (token.contents == 'else') {
            nodelist_false = parser.parse(['endif']);
            parser.delete_first_token();
        } else {
            nodelist_false = new node.NodeList();
        }
        return new IfNode(val, nodelist_true, nodelist_false);
    },

    'autoescape' : function(parser, token) {
        // Force autoescape behaviour for this block.
        var args = token.split_contents()

        if (args.length != 2) 
            throw new Thistle.TemplateSyntaxError("'autoescape' tag requires exactly one argument.")

        var arg = args[1];
        if (arg != 'on' && arg != 'off')
            throw new Thistle.TemplateSyntaxError("'autoescape' argument should be 'on' or 'off'")
        
        var nodelist = parser.parse(['endautoescape']);
        parser.delete_first_token();
        return new AutoEscapeControlNode((arg == 'on'), nodelist);
    },

    'filter' : function(parser, token) {
        // Force autoescape behaviour for this block.
        
        var rest = token.split_contents().slice(1).join(' ');
        var filter_expr = parser.compile_filter("var|" + rest);

        nodelist = parser.parse(['endfilter']);
        parser.delete_first_token();
        return new FilterNode(filter_expr, nodelist);
    },
};

module.exports = DefaultTags;

File: py/thistle/default_tags.py
import re
from .library import Library
from .node import Node, NodeList
from thistle import TemplateSyntaxError, VariableDoesNotExist
from .safestring import mark_safe
from .smartif import IfParser, Literal
from itertools import cycle as itertools_cycle

register = Library()

class AutoEscapeControlNode(Node):
    """Implements the actions of the autoescape tag."""
    def __init__(self, setting, nodelist):
        self.setting, self.nodelist = setting, nodelist

    def render(self, context):
        old_setting = context.autoescape
        context.autoescape = self.setting
        output = self.nodelist.render(context)
        context.autoescape = old_setting
        if self.setting:
            return mark_safe(output)
        else:
            return output

class CommentNode(Node):
    def render(self, context):
        return ''

class CycleNode(Node):
    def __init__(self, cyclevars, variable_name=None, silent=False):
        self.cyclevars = cyclevars
        self.variable_name = variable_name
        self.silent = silent

    def render(self, context):
        if self not in context.render_context:
            # First time the node is rendered in template
            context.render_context[self] = itertools_cycle(self.cyclevars)
        cycle_iter = context.render_context[self]
        value = cycle_iter.next().resolve(context)
        if self.variable_name:
            context[self.variable_name] = value
        if self.silent:
            return ''
        return value

class FilterNode(Node):
    def __init__(self, filter_expr, nodelist):
        self.filter_expr, self.nodelist = filter_expr, nodelist

    def render(self, context):
        output = self.nodelist.render(context)
        # Apply filters.
        context.update({'var': output})
        filtered = self.filter_expr.resolve(context)
        context.pop()
        return filtered

class ForNode(Node):
    child_nodelists = ('nodelist_loop', 'nodelist_empty')

    def __init__(self, loopvars, sequence, is_reversed, nodelist_loop, nodelist_empty=None):
        self.loopvars, self.sequence = loopvars, sequence
        self.is_reversed = is_reversed
        self.nodelist_loop = nodelist_loop
        if nodelist_empty is None:
            self.nodelist_empty = NodeList()
        else:
            self.nodelist_empty = nodelist_empty

    def __repr__(self):
        reversed_text = self.is_reversed and ' reversed' or ''
        return "<For Node: for %s in %s, tail_len: %d%s>" % \
            (', '.join(self.loopvars), self.sequence, len(self.nodelist_loop),
             reversed_text)

    def __iter__(self):
        for node in self.nodelist_loop:
            yield node
        for node in self.nodelist_empty:
            yield node

    def render(self, context):
        if 'forloop' in context:
            parentloop = context['forloop']
        else:
            parentloop = {}
        context.push()
        try:
            values = self.sequence.resolve(context, True)
        except VariableDoesNotExist:
            values = []
        if values is None:
            values = []
        if not hasattr(values, '__len__'):
            values = list(values)
        len_values = len(values)
        if len_values < 1:
            context.pop()
            return self.nodelist_empty.render(context)
        nodelist = NodeList()
        if self.is_reversed:
            values = reversed(values)
        unpack = len(self.loopvars) > 1
        # Create a forloop value in the context.  We'll update counters on each
        # iteration just below.
        loop_dict = context['forloop'] = {'parentloop': parentloop}
        for i, item in enumerate(values):
            # Shortcuts for current loop iteration number.
            loop_dict['counter0'] = i
            loop_dict['counter'] = i+1
            # Reverse counter iteration numbers.
            loop_dict['revcounter'] = len_values - i
            loop_dict['revcounter0'] = len_values - i - 1
            # Boolean values designating first and last times through loop.
            loop_dict['first'] = (i == 0)
            loop_dict['last'] = (i == len_values - 1)

            pop_context = False
            if unpack:
                # If there are multiple loop variables, unpack the item into
                # them.
                try:
                    unpacked_vars = dict(zip(self.loopvars, item))
                except TypeError:
                    pass
                else:
                    pop_context = True
                    context.update(unpacked_vars)
            else:
                context[self.loopvars[0]] = item
            for node in self.nodelist_loop:
                nodelist.append(node.render(context))
            if pop_context:
                # The loop variables were pushed on to the context so pop them
                # off again. This is necessary because the tag lets the length
                # of loopvars differ to the length of each set of items and we
                # don't want to leave any vars from the previous loop on the
                # context.
                context.pop()
        context.pop()
        return nodelist.render(context)
class IfNode(Node):
    child_nodelists = ('nodelist_true', 'nodelist_false')

    def __init__(self, var, nodelist_true, nodelist_false=None):
        self.nodelist_true, self.nodelist_false = nodelist_true, nodelist_false
        self.var = var

    def __repr__(self):
        return "<If node>"

    def __iter__(self):
        for node in self.nodelist_true:
            yield node
        for node in self.nodelist_false:
            yield node

    def render(self, context):
        try:
            var = self.var.eval(context)
        except VariableDoesNotExist:
            var = None

        if var:
            return self.nodelist_true.render(context)
        else:
            return self.nodelist_false.render(context)

class SpacelessNode(Node):
    def __init__(self, nodelist):
        self.nodelist = nodelist

    def render(self, context):
        """Returns the given HTML with spaces between tags removed."""
        value = self.nodelist.render(context).strip()
        return re.sub(r'>\s+<', '><', value)

class TemplateLiteral(Literal):
    def __init__(self, value, text):
        self.value = value
        self.text = text # for better error messages

    def display(self):
        return self.text

    def eval(self, context):
        return self.value.resolve(context, ignore_failures=True)

class TemplateIfParser(IfParser):
    error_class = TemplateSyntaxError

    def __init__(self, parser, *args, **kwargs):
        self.template_parser = parser
        super(TemplateIfParser, self).__init__(*args, **kwargs)

    def create_var(self, value):
        return TemplateLiteral(self.template_parser.compile_filter(value), value)

#
#
#
@register.tag
def autoescape(parser, token):
    """
    Force autoescape behaviour for this block.
    """
    args = token.contents.split()
    if len(args) != 2:
        raise TemplateSyntaxError("'autoescape' tag requires exactly one argument.")
    arg = args[1]
    if arg not in (u'on', u'off'):
        raise TemplateSyntaxError("'autoescape' argument should be 'on' or 'off'")
    nodelist = parser.parse(('endautoescape',))
    parser.delete_first_token()
    return AutoEscapeControlNode((arg == 'on'), nodelist)

@register.tag
def comment(parser, token):
    """
    Ignores everything between ``{% comment %}`` and ``{% endcomment %}``.
    """
    parser.skip_past('endcomment')
    return CommentNode()


@register.tag
def cycle(parser, token):
    """
    Cycles among the given strings each time this tag is encountered.

    Within a loop, cycles among the given strings each time through
    the loop::

        {% for o in some_list %}
            <tr class="{% cycle 'row1' 'row2' %}">
                ...
            </tr>
        {% endfor %}

    Outside of a loop, give the values a unique name the first time you call
    it, then use that name each sucessive time through::

            <tr class="{% cycle 'row1' 'row2' 'row3' as rowcolors %}">...</tr>
            <tr class="{% cycle rowcolors %}">...</tr>
            <tr class="{% cycle rowcolors %}">...</tr>

    You can use any number of values, separated by spaces. Commas can also
    be used to separate values; if a comma is used, the cycle values are
    interpreted as literal strings.

    The optional flag "silent" can be used to prevent the cycle declaration
    from returning any value::

        {% cycle 'row1' 'row2' as rowcolors silent %}{# no value here #}
        {% for o in some_list %}
            <tr class="{% cycle rowcolors %}">{# first value will be "row1" #}
                ...
            </tr>
        {% endfor %}

    """

    # Note: This returns the exact same node on each {% cycle name %} call;
    # that is, the node object returned from {% cycle a b c as name %} and the
    # one returned from {% cycle name %} are the exact same object. This
    # shouldn't cause problems (heh), but if it does, now you know.
    #
    # Ugly hack warning: This stuffs the named template dict into parser so
    # that names are only unique within each template (as opposed to using
    # a global variable, which would make cycle names have to be unique across
    # *all* templates.

    args = token.split_contents()

    if len(args) < 2:
        raise TemplateSyntaxError("'cycle' tag requires at least two arguments")

    if ',' in args[1]:
        # Backwards compatibility: {% cycle a,b %} or {% cycle a,b as foo %}
        # case.
        args[1:2] = ['"%s"' % arg for arg in args[1].split(",")]

    if len(args) == 2:
        # {% cycle foo %} case.
        name = args[1]
        if not hasattr(parser, '_namedCycleNodes'):
            raise TemplateSyntaxError("No named cycles in template. '%s' is not defined" % name)
        if not name in parser._namedCycleNodes:
            raise TemplateSyntaxError("Named cycle '%s' does not exist" % name)
        return parser._namedCycleNodes[name]

    as_form = False

    if len(args) > 4:
        # {% cycle ... as foo [silent] %} case.
        if args[-3] == "as":
            if args[-1] != "silent":
                raise TemplateSyntaxError("Only 'silent' flag is allowed after cycle's name, not '%s'." % args[-1])
            as_form = True
            silent = True
            args = args[:-1]
        elif args[-2] == "as":
            as_form = True
            silent = False

    if as_form:
        name = args[-1]
        values = [parser.compile_filter(arg) for arg in args[1:-2]]
        node = CycleNode(values, name, silent=silent)
        if not hasattr(parser, '_namedCycleNodes'):
            parser._namedCycleNodes = {}
        parser._namedCycleNodes[name] = node
    else:
        values = [parser.compile_filter(arg) for arg in args[1:]]
        node = CycleNode(values)
    return node

@register.tag('if')
def do_if(parser, token):
    """
    The ``{% if %}`` tag evaluates a variable, and if that variable is "true"
    (i.e., exists, is not empty, and is not a false boolean value), the
    contents of the block are output:

    ::

        {% if athlete_list %}
            Number of athletes: {{ athlete_list|count }}
        {% else %}
            No athletes.
        {% endif %}

    In the above, if ``athlete_list`` is not empty, the number of athletes will
    be displayed by the ``{{ athlete_list|count }}`` variable.

    As you can see, the ``if`` tag can take an option ``{% else %}`` clause
    that will be displayed if the test fails.

    ``if`` tags may use ``or``, ``and`` or ``not`` to test a number of
    variables or to negate a given variable::

        {% if not athlete_list %}
            There are no athletes.
        {% endif %}

        {% if athlete_list or coach_list %}
            There are some athletes or some coaches.
        {% endif %}

        {% if athlete_list and coach_list %}
            Both atheletes and coaches are available.
        {% endif %}

        {% if not athlete_list or coach_list %}
            There are no athletes, or there are some coaches.
        {% endif %}

        {% if athlete_list and not coach_list %}
            There are some athletes and absolutely no coaches.
        {% endif %}

    Comparison operators are also available, and the use of filters is also
    allowed, for example::

        {% if articles|length >= 5 %}...{% endif %}

    Arguments and operators _must_ have a space between them, so
    ``{% if 1>2 %}`` is not a valid if tag.

    All supported operators are: ``or``, ``and``, ``in``, ``not in``
    ``==`` (or ``=``), ``!=``, ``>``, ``>=``, ``<`` and ``<=``.

    Operator precedence follows Python.
    """
    bits = token.split_contents()[1:]
    var = TemplateIfParser(parser, bits).parse()
    nodelist_true = parser.parse(('else', 'endif'))
    token = parser.next_token()
    if token.contents == 'else':
        nodelist_false = parser.parse(('endif',))
        parser.delete_first_token()
    else:
        nodelist_false = NodeList()
    return IfNode(var, nodelist_true, nodelist_false)

@register.tag('filter')
def do_filter(parser, token):
    """
    Filters the contents of the block through variable filters.

    Filters can also be piped through each other, and they can have
    arguments -- just like in variable syntax.

    Sample usage::

        {% filter force_escape|lower %}
            This text will be HTML-escaped, and will appear in lowercase.
        {% endfilter %}
    """
    _, rest = token.contents.split(None, 1)
    filter_expr = parser.compile_filter("var|%s" % (rest))
    for func, unused in filter_expr.filters:
        if getattr(func, '_decorated_function', func).__name__ in ('escape', 'safe'):
            raise TemplateSyntaxError('"filter %s" is not permitted.  Use the "autoescape" tag instead.' % func.__name__)
    nodelist = parser.parse(('endfilter',))
    parser.delete_first_token()
    return FilterNode(filter_expr, nodelist)

@register.tag('for')
def do_for(parser, token):
    """
    Loops over each item in an array.

    For example, to display a list of athletes given ``athlete_list``::

        <ul>
        {% for athlete in athlete_list %}
            <li>{{ athlete.name }}</li>
        {% endfor %}
        </ul>

    You can loop over a list in reverse by using
    ``{% for obj in list reversed %}``.

    You can also unpack multiple values from a two-dimensional array::

        {% for key,value in dict.items %}
            {{ key }}: {{ value }}
        {% endfor %}

    The ``for`` tag can take an optional ``{% empty %}`` clause that will
    be displayed if the given array is empty or could not be found::

        <ul>
          {% for athlete in athlete_list %}
            <li>{{ athlete.name }}</li>
          {% empty %}
            <li>Sorry, no athletes in this list.</li>
          {% endfor %}
        <ul>

    The above is equivalent to -- but shorter, cleaner, and possibly faster
    than -- the following::

        <ul>
          {% if althete_list %}
            {% for athlete in athlete_list %}
              <li>{{ athlete.name }}</li>
            {% endfor %}
          {% else %}
            <li>Sorry, no athletes in this list.</li>
          {% endif %}
        </ul>

    The for loop sets a number of variables available within the loop:

        ==========================  ================================================
        Variable                    Description
        ==========================  ================================================
        ``forloop.counter``         The current iteration of the loop (1-indexed)
        ``forloop.counter0``        The current iteration of the loop (0-indexed)
        ``forloop.revcounter``      The number of iterations from the end of the
                                    loop (1-indexed)
        ``forloop.revcounter0``     The number of iterations from the end of the
                                    loop (0-indexed)
        ``forloop.first``           True if this is the first time through the loop
        ``forloop.last``            True if this is the last time through the loop
        ``forloop.parentloop``      For nested loops, this is the loop "above" the
                                    current one
        ==========================  ================================================

    """
    bits = token.contents.split()
    if len(bits) < 4:
        raise TemplateSyntaxError("'for' statements should have at least four"
                                  " words: %s" % token.contents)

    is_reversed = bits[-1] == 'reversed'
    in_index = is_reversed and -3 or -2
    if bits[in_index] != 'in':
        raise TemplateSyntaxError("'for' statements should use the format"
                                  " 'for x in y': %s" % token.contents)

    loopvars = re.split(r' *, *', ' '.join(bits[1:in_index]))
    for var in loopvars:
        if not var or ' ' in var:
            raise TemplateSyntaxError("'for' tag received an invalid argument:"
                                      " %s" % token.contents)

    sequence = parser.compile_filter(bits[in_index+1])
    nodelist_loop = parser.parse(('empty', 'endfor',))
    token = parser.next_token()
    if token.contents == 'empty':
        nodelist_empty = parser.parse(('endfor',))
        parser.delete_first_token()
    else:
        nodelist_empty = None
    return ForNode(loopvars, sequence, is_reversed, nodelist_loop, nodelist_empty)

@register.tag
def spaceless(parser, token):
    """
    Removes whitespace between HTML tags, including tab and newline characters.

    Example usage::

        {% spaceless %}
            <p>
                <a href="foo/">Foo</a>
            </p>
        {% endspaceless %}

    This example would return this HTML::

        <p><a href="foo/">Foo</a></p>

    Only space between *tags* is normalized -- not space between tags and text.
    In this example, the space around ``Hello`` won't be stripped::

        {% spaceless %}
            <strong>
                Hello
            </strong>
        {% endspaceless %}
    """
    nodelist = parser.parse(('endspaceless',))
    parser.delete_first_token()
    return SpacelessNode(nodelist)
